### ISSUE TITLE
Try install without depext at least once

### ISSUE THREAD

**Issue by toots** [ocaml/opam#4860]
We recently had to [fix a depext on libopus](https://github.com/ocaml/opam-repository/pull/19680) and noticed that `opam` `2.1.0` will keep failing even if the depext is installed with the fixed name. This is not really robust as depext do change over OS release. Maybe `--assume-depext` should be implicitly tried at least once?

**Comment by toots** [comment]
No, `conf-libopus` was referring to the wrong OS package but the local system had the right depext already installed.

**Comment by dra27** [comment]
I think you 'dwant `--no-depexts` here? If the depext has changed name, then `--assume-depext` wouldn't be enough (because opam would still be checking to see if that package exists)?

I don't think we should automatically do things which wallpaper over issues on error - it makes opam _more_ surprising. In this case, it also makes it slower - opam's always going to be checking for a non-existent depext and then carrying on until someone is annoyed enough to investigate why and fix the actual issue.

We'll discuss it further, but I think there are three things we could/should do instead:
- If it's not already, the error message when failing to install should clearly mention the `--no-depexts` option (but I think it does?)
- if opam is in interactive mode then it _could_ offer to re-run with `--no-depexts` (but as I said, I don't think this should be automated - the underlying problem here is that the metadata are wrong, and the metadata should be fixed).
- The testing carried out on opam-repository is always expanding - we're on the cusp of doing regular health-check runs on macOS (and then Windows), so I would hope that we'll be able to catch these depext changes either just as, or even before, our users

**Comment by toots** [comment]
I understand all that but, at the end of the day, there are multiple ways of installing external dependencies. Compiling manually, installing via a package manager, potentially one of many.

On macos, typically, one can install via `brew` or `macports`.

Also, some dependencies can have different package names, for instance `libsrt` on debian has two variants, one with `ssl` and one with `gnutls`.

You also have cases where two copies of the same external dependencies are installed, for instance one for `x86` and one for `x64` or a cross-compiled one.

There is no way of knowing in advance which package manager and package name (if any) should be used for any given situation.

It is _really good_ that `opam` has support for external dependencies, it is an incredible option and makes everything so easy to use, no doubt about it.

However, one should also think about usability from the user perspective and not tie the user to a specific package manager and/or a specific package name.

The most basic situation is to try and install a OCaml package without making assumption about external dependencies. Then, if that doesn't work, it seems okay to make an educated guess and try installing an external dependency.

Even better would be if, at this stage, an interactive question is prompted to the user:
> `opam` is not able to install your package at the moment. It might be that you need to install the following dependencies: .... Would you like us to do it for you?

That would be a great user experience.

**Comment by dbuenzli** [comment]
Some other communities like coq use it like, so I would rather say it is a developer tool. I don't think it's suitable to install applications, especially for people who are not developers at all.

I think the best experience for your users is to provide them the ability to install liquidsoap with their system package manager.

**Comment by dbuenzli** [comment]
Did you have a look into things like https://github.com/AltGr/opam-bundle ?

**Comment by toots** [comment]
> Some other communities like coq use it like, so I would rather say it is a developer tool. I don't think it's suitable to install applications, especially for people who are not developers at all.
> 
> I think the best experience for your users is to provide them the ability to install liquidsoap with their system package manager. (Having to learn something entirely new just to be able to install liquidsoap is a barrier to your software)

I have packaged for Debian for years and was involved closely with the OCaml packaging there so I can say with some degree of confidence that system level packaging is a poor solution to help users install complex OCaml applications. 

`opam` has proven to be a fantastic tool. It handles the complexity of:
* Resolving ocaml module dependencies and upgrading them to the latest compatible version while rebuilding all transitive dependencies
* Installing optional dependencies and supporting optional features
* Installing external dependencies

Going back to system-level packaging for each of these features is a step back in user experience and also in developers burden since this means packaging, updating and rebuilding every single ocaml dependency and their dependency for each different OS-specific package manager.

That's not to say that we don't do it, we provide Debian, Ubuntu, Docker and Alpine packages for a plethora of version and architectures but, when it comes to giving users a single, unified way of installing the application with the latest ocaml dependencies and optional features, `opam` clearly provides a much, much better user experience.

At the end of the day, I think @dra27 and myself are not far appart, we both want `opam` to do what it's told by default, my point is that, to be more inclusive, I believe we should assume that the user installing a `opam` package has no idea about external depencies and should prompted about what `opam` is doing when handling so.

Otherwise, this keels akind to having your tech savvy friend seating next to you telling you: "don't worry dude, let me handle this", leaving you with a total sense of confusion about what just happened.

I believe that we should take all opportunities to explain what is being done internally when specific decisions or assumptions are being made by any software, give the user a friendly, explanatory message about what and why this is being done and use this opportunity to give more clarity and give the user a chance to learn and understand better what is going on.
